malsim/mal_simulator.py

```diff
@@ -349,19 +345,23 @@ def _attacker_step(
                 agent.reward += node_reward
                 for d_agent in self._get_defender_agents():
                     d_agent.reward -= node_reward
-                enabled_nodes.append(node)
+                compromised_nodes.add(node)
 
                 logger.info(
                     'Attacker agent "%s" compromised "%s"(%d).',
                     agent.name, node.full_name, node.id
                 )
-
-                # Update attacker action surface
-                query.update_attack_surface_add_nodes(
-                    attacker, agent.action_surface, [node])
             else:
                 logger.warning("Attacker could not compromise %s",
-                            node.full_name)
+                               node.full_name)
+
+        # Update attacker action surface
+        new_attack_surface = query.calculate_attack_surface(
+            attacker, from_nodes=compromised_nodes, skip_compromised=True
+        )
+
+        # TODO: why extend?
+        agent.action_surface |= new_attack_surface
```

Comment:
What do you mean by why extend? This is just determining the attack surface from the node it just compromised. The old nodes in the attack surface are still potential options?

Reply:
Right, it's a leftover reminder to discuss this. I will remove it.

Reply:
Actually, the discussion is still to be done right? If the simulator should include the already compromised nodes in the attack surface sent to the attacker or not, which is sth the attacker knows already. If the surface is huge and the attacker needs to extract only the non-compromised ones, it makes it slow or at least more complicated than it needs to be. Perhaps the state could include the non-compromised nodes in the attack surface already?

Reply:
I think we might be combining two related discussions into one. This does not explicitly have anything to do with compromised nodes . It just means that the old values in the attack surface should still be part of it, whether some of them have already been compromised or not is irrelevant here.

I do remember/understand the point you were making that the attack surface scales with the size of the attack graph, which it absolutely does, as do the `reached_attack_steps` on the `AttackGraph` `Attacker`. And that we may wish to have optimized versions that just provide deltas and let the agents handle them as they see fit.

We may want this configurable. But there's no doubt that right now it should keep the previous entries in the attack surface, it replaces the old update method.

Reply:
I tried playing with the code a bit, and returning all possible actions in the attack surface makes sense, even compromised ones. The operation to extract the uncompromised steps from the attack surface will be the same regardless of whether it happens in the agent or the simulator. Since some agents may not need this extraction, it's better left to the agents' side to occur on demand.